Travis unittest failures:

* https://travis-ci.org/web-platform-tests/wpt/jobs/452825483
* https://travis-ci.org/web-platform-tests/wpt/jobs/452825484

```
_________ ERROR collecting wptserve/tests/functional/test_handlers.py __________
ImportError while importing test module '/home/travis/build/web-platform-tests/wpt/tools/wptserve/tests/functional/test_handlers.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
.tox/py27/lib/python2.7/site-packages/six.py:709: in exec_
    exec("""exec _code_ in _globs_, _locs_""")
wptserve/tests/functional/test_handlers.py:14: in <module>
    from serve.serve import WorkersHandler, WindowHandler, AnyHtmlHandler,\
E   ImportError: No module named serve
```
and
```
ERROR: InvocationError for command '/home/travis/build/web-platform-tests/wpt/tools/.tox/py36/bin/pytest --cov' (exited with code 2)
py36-flake8 create: /home/travis/build/web-platform-tests/wpt/tools/.tox/py36-flake8
py36-flake8 installdeps: -rrequirements_flake8.txt
py36-flake8 installed: flake8==3.6.0,flake8-polyfill==1.0.2,mccabe==0.6.1,pep8-naming==0.7.0,pycodestyle==2.4.0,pyflakes==2.0.0
py36-flake8 run-test-pre: PYTHONHASHSEED='3166492845'
py36-flake8 runtests: commands[0] | flake8 --append-config=py36-flake8.ini
___________________________________ summary ____________________________________
ERROR:   py36: commands failed
```

--- a/tools/wptserve/tests/functional/base.py
+++ b/tools/wptserve/tests/functional/base.py
@@ -117,10 +117,10 @@ class TestWrapperHandlerUsingServer(TestUsingServer):
     remove it in tearDown.'''
     dummy_js_files = {}
 
-    def gen_js_file(self, filename, empty=True, content=''):
+    def gen_js_file(self, filename, empty=True, content=b''):
         self.remove_js_file(filename)
 
-        with open(filename, 'w') as fp:
+        with open(filename, 'wb') as fp:
             if not empty:
                 fp.write(content)
 
@@ -146,7 +146,7 @@ class TestWrapperHandlerUsingServer(TestUsingServer):
         self.assertEqual(200, resp.getcode())
         self.assertEqual(header_data, resp.info()['Content-Type'])
 
-        with open(os.path.join(doc_root, req_file), 'r') as fp:
+        with open(os.path.join(doc_root, req_file), 'rb') as fp:
             self.assertEqual(fp.read(), resp.read())
 
     def tearDown(self):

--- a/tools/wptserve/tests/functional/base.py
+++ b/tools/wptserve/tests/functional/base.py
@@ -108,3 +108,50 @@ class TestUsingH2Server:
     def teardown_method(self, test_method):
         self.conn.close()
         self.server.stop()
+
+
+class TestWrapperHandlerUsingServer(TestUsingServer):
+    '''For a wrapper handler, a .js dummy testing file is requried to render
+    the html file. This class extends the TestUsingServer and do some some
+    extra work: it tries to generate the dummy .js file in setUp and
+    remove it in tearDown.'''
+    dummy_js_files = {}
+
+    def gen_js_file(self, filename, empty=True, content=''):
+        self.remove_js_file(filename)
+
+        with open(filename, 'w') as fp:
+            if not empty:
+                fp.write(content)
+
+    def remove_js_file(self, filename):
+        if os.path.exists(filename):
+            os.remove(filename)
+
+    def setUp(self):
+        super(TestWrapperHandlerUsingServer, self).setUp()
+
+        for filename, content in self.dummy_js_files.items():
+            filepath = os.path.join(doc_root, filename)
+            if content == '':
+                self.gen_js_file(filepath)
+            else:
+                self.gen_js_file(filepath, False, content)
+
+    def run_wrapper_test(self, req_file, header_data, wrapper_handler):
+        route = ('GET', req_file, wrapper_handler())
+        self.server.router.register(*route)
+
+        resp = self.request(route[1])
+        self.assertEqual(200, resp.getcode())
+        self.assertEqual(header_data, resp.info()['Content-Type'])
+
+        with open(os.path.join(doc_root, req_file), 'r') as fp:
+            self.assertEqual(fp.read(), resp.read())
+
+    def tearDown(self):
+        super(TestWrapperHandlerUsingServer, self).tearDown()
+
+        for filename, _ in self.dummy_js_files.items():
+            filepath = os.path.join(doc_root, filename)
+            self.remove_js_file(filepath)

--- a/tools/wptserve/tests/functional/docroot/bar.any.worker.js
+++ b/tools/wptserve/tests/functional/docroot/bar.any.worker.js
@@ -1,0 +1,9 @@
+
+self.GLOBAL = {
+  isWindow: function() { return false; },
+  isWorker: function() { return true; },
+};
+importScripts("/resources/testharness.js");
+
+importScripts("/bar.any.js");
+done();

--- a/tools/wptserve/tests/functional/docroot/foo.any.html
+++ b/tools/wptserve/tests/functional/docroot/foo.any.html
@@ -1,0 +1,14 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script>
+self.GLOBAL = {
+  isWindow: function() { return true; },
+  isWorker: function() { return false; },
+};
+</script>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+
+<div id=log></div>
+<script src="/foo.any.js"></script>

--- a/tools/wptserve/tests/functional/docroot/foo.any.serviceworker.html
+++ b/tools/wptserve/tests/functional/docroot/foo.any.serviceworker.html
@@ -1,0 +1,15 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+<script>
+(async function() {
+  const scope = 'does/not/exist';
+  let reg = await navigator.serviceWorker.getRegistration(scope);
+  if (reg) await reg.unregister();
+  reg = await navigator.serviceWorker.register("/foo.any.worker.js", {scope});
+  fetch_tests_from_worker(reg.installing);
+})();
+</script>

--- a/tools/wptserve/tests/functional/docroot/foo.any.sharedworker.html
+++ b/tools/wptserve/tests/functional/docroot/foo.any.sharedworker.html
@@ -1,0 +1,9 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+<script>
+fetch_tests_from_worker(new SharedWorker("/foo.any.worker.js"));
+</script>

--- a/tools/wptserve/tests/functional/docroot/foo.any.worker.html
+++ b/tools/wptserve/tests/functional/docroot/foo.any.worker.html
@@ -1,0 +1,9 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+<script>
+fetch_tests_from_worker(new Worker("/foo.any.worker.js"));
+</script>

--- a/tools/wptserve/tests/functional/docroot/foo.window.html
+++ b/tools/wptserve/tests/functional/docroot/foo.window.html
@@ -1,0 +1,8 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+
+<div id=log></div>
+<script src="/foo.window.js"></script>

--- a/tools/wptserve/tests/functional/docroot/foo.worker.html
+++ b/tools/wptserve/tests/functional/docroot/foo.worker.html
@@ -1,0 +1,9 @@
+<!doctype html>
+<meta charset=utf-8>
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+<script>
+fetch_tests_from_worker(new Worker("/foo.worker.js"));
+</script>

--- a/tools/wptserve/tests/functional/test_handlers.py
+++ b/tools/wptserve/tests/functional/test_handlers.py
@@ -8,8 +8,11 @@ import pytest
 from six.moves.urllib.error import HTTPError
 
 wptserve = pytest.importorskip("wptserve")
-from .base import TestUsingServer, TestUsingH2Server, doc_root
+from .base import TestUsingServer, TestUsingH2Server, \
+    TestWrapperHandlerUsingServer, doc_root
 
+from serve.serve import WorkersHandler, WindowHandler, AnyHtmlHandler,\
+    SharedWorkersHandler, ServiceWorkersHandler, AnyWorkerHandler
 
 class TestFileHandler(TestUsingServer):
     def test_GET(self):
@@ -386,6 +389,65 @@ class TestH2Handler(TestUsingH2Server):
         assert resp.status == 203
         assert resp.headers['test'][0] == 'passed'
         assert resp.read() == ''
+
+
+class TestWorkersHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'foo.worker.js': '',
+                      'foo.any.js': ''}
+
+    def test_any_worker_html(self):
+        self.run_wrapper_test('foo.any.worker.html',
+                              'text/html', WorkersHandler)
+
+    def test_worker_html(self):
+        self.run_wrapper_test('foo.worker.html',
+                              'text/html', WorkersHandler)
+
+
+class TestWindowHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'foo.window.js': ''}
+
+    def test_window_html(self):
+        self.run_wrapper_test('foo.window.html',
+                              'text/html', WindowHandler)
+
+
+class TestAnyHtmlHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'foo.any.js': ''}
+
+    def test_any_html(self):
+        self.run_wrapper_test('foo.any.html',
+                              'text/html', AnyHtmlHandler)
+
+
+class TestSharedWorkersHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'foo.any.js': '// META: global=sharedworker\n'}
+
+    def test_any_sharedworkers_html(self):
+        self.run_wrapper_test('foo.any.sharedworker.html',
+                              'text/html', SharedWorkersHandler)
+
+
+class TestServiceWorkersHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'foo.any.js': '// META: global=serviceworker\n'}
+
+    def test_serviceworker_html(self):
+        self.run_wrapper_test('foo.any.serviceworker.html',
+                              'text/html', ServiceWorkersHandler)
+
+
+class TestAnyWorkerHandler(TestWrapperHandlerUsingServer):
+
+    dummy_js_files = {'bar.any.js': ''}
+
+    def test_any_work_js(self):
+        self.run_wrapper_test('bar.any.worker.js', 'text/javascript',
+                              AnyWorkerHandler)
 
 
 if __name__ == '__main__':

--- a/tools/wptserve/tests/functional/test_handlers.py
+++ b/tools/wptserve/tests/functional/test_handlers.py
@@ -8,11 +8,10 @@ import pytest
 from six.moves.urllib.error import HTTPError
 
 wptserve = pytest.importorskip("wptserve")
-from .base import TestUsingServer, TestUsingH2Server, \
-    TestWrapperHandlerUsingServer, doc_root
+from .base import TestUsingServer, TestUsingH2Server, doc_root
+from .base import TestWrapperHandlerUsingServer
 
-from serve.serve import WorkersHandler, WindowHandler, AnyHtmlHandler,\
-    SharedWorkersHandler, ServiceWorkersHandler, AnyWorkerHandler
+from serve import serve
 
 class TestFileHandler(TestUsingServer):
     def test_GET(self):
@@ -392,62 +391,56 @@ class TestH2Handler(TestUsingH2Server):
 
 
 class TestWorkersHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'foo.worker.js': '',
                       'foo.any.js': ''}
 
     def test_any_worker_html(self):
         self.run_wrapper_test('foo.any.worker.html',
-                              'text/html', WorkersHandler)
+                              'text/html', serve.WorkersHandler)
 
     def test_worker_html(self):
         self.run_wrapper_test('foo.worker.html',
-                              'text/html', WorkersHandler)
+                              'text/html', serve.WorkersHandler)
 
 
 class TestWindowHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'foo.window.js': ''}
 
     def test_window_html(self):
         self.run_wrapper_test('foo.window.html',
-                              'text/html', WindowHandler)
+                              'text/html', serve.WindowHandler)
 
 
 class TestAnyHtmlHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'foo.any.js': ''}
 
     def test_any_html(self):
         self.run_wrapper_test('foo.any.html',
-                              'text/html', AnyHtmlHandler)
+                              'text/html', serve.AnyHtmlHandler)
 
 
 class TestSharedWorkersHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'foo.any.js': '// META: global=sharedworker\n'}
 
     def test_any_sharedworkers_html(self):
         self.run_wrapper_test('foo.any.sharedworker.html',
-                              'text/html', SharedWorkersHandler)
+                              'text/html', serve.SharedWorkersHandler)
 
 
 class TestServiceWorkersHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'foo.any.js': '// META: global=serviceworker\n'}
 
     def test_serviceworker_html(self):
         self.run_wrapper_test('foo.any.serviceworker.html',
-                              'text/html', ServiceWorkersHandler)
+                              'text/html', serve.ServiceWorkersHandler)
 
 
 class TestAnyWorkerHandler(TestWrapperHandlerUsingServer):
-
     dummy_js_files = {'bar.any.js': ''}
 
     def test_any_work_js(self):
         self.run_wrapper_test('bar.any.worker.js', 'text/javascript',
-                              AnyWorkerHandler)
+                              serve.AnyWorkerHandler)
 
 
 if __name__ == '__main__':

--- a/tools/wptserve/tests/functional/test_handlers.py
+++ b/tools/wptserve/tests/functional/test_handlers.py
@@ -391,8 +391,8 @@ class TestH2Handler(TestUsingH2Server):
 
 
 class TestWorkersHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'foo.worker.js': '',
-                      'foo.any.js': ''}
+    dummy_js_files = {'foo.worker.js': b'',
+                      'foo.any.js': b''}
 
     def test_any_worker_html(self):
         self.run_wrapper_test('foo.any.worker.html',
@@ -404,7 +404,7 @@ class TestWorkersHandler(TestWrapperHandlerUsingServer):
 
 
 class TestWindowHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'foo.window.js': ''}
+    dummy_js_files = {'foo.window.js': b''}
 
     def test_window_html(self):
         self.run_wrapper_test('foo.window.html',
@@ -412,7 +412,7 @@ class TestWindowHandler(TestWrapperHandlerUsingServer):
 
 
 class TestAnyHtmlHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'foo.any.js': ''}
+    dummy_js_files = {'foo.any.js': b''}
 
     def test_any_html(self):
         self.run_wrapper_test('foo.any.html',
@@ -420,7 +420,7 @@ class TestAnyHtmlHandler(TestWrapperHandlerUsingServer):
 
 
 class TestSharedWorkersHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'foo.any.js': '// META: global=sharedworker\n'}
+    dummy_js_files = {'foo.any.js': b'// META: global=sharedworker\n'}
 
     def test_any_sharedworkers_html(self):
         self.run_wrapper_test('foo.any.sharedworker.html',
@@ -428,7 +428,7 @@ class TestSharedWorkersHandler(TestWrapperHandlerUsingServer):
 
 
 class TestServiceWorkersHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'foo.any.js': '// META: global=serviceworker\n'}
+    dummy_js_files = {'foo.any.js': b'// META: global=serviceworker\n'}
 
     def test_serviceworker_html(self):
         self.run_wrapper_test('foo.any.serviceworker.html',
@@ -436,7 +436,7 @@ class TestServiceWorkersHandler(TestWrapperHandlerUsingServer):
 
 
 class TestAnyWorkerHandler(TestWrapperHandlerUsingServer):
-    dummy_js_files = {'bar.any.js': ''}
+    dummy_js_files = {'bar.any.js': b''}
 
     def test_any_work_js(self):
         self.run_wrapper_test('bar.any.worker.js', 'text/javascript',

